src/dag.ts

```diff
@@ -71,19 +71,20 @@ export function iterate<T>(nodes: INode<T>[], edges: IEdge[],
     return unvisited === false ? childNodes : childNodes.filter(filterNotVisited);
   }
 
+  // Initialize the stacks with all the nodes that have no parents.
   Array.prototype.push.apply(nodeStack, getFirstNodes());
 
   while (nodeStack.length) {
-    // Take a layer from the stack
-    const node = nodeStack.pop();
+    // Take the top layer from the stack
+    const node = nodeStack.shift();
```

Comment:
question, should the node be taken out from the beginning instead of the end? Since the child is pushed to the end of the array?

Reply:
I implemented the iterator as depth-first graph search using a stack; this change would turn `nodeStack` into a queue and hence implement breadth-first search. BFS seems reasonable for drawing the DAG, however DFS is more logical for computation and optimization (e.g. merging together 2 consecutive operations, such as Conv and Pool). I think ultimately this library should support both ways, maybe as `iterateBfs` and `iterateDfs`.

Let's keep it a stack for now, and add the new functions afterwards; I will add it tonight together with tests a new minor release.

Reply:
sounds good thanks.